Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,172 @@
+    public static func bootstrapSystemMetrics(pollInterval interval: DispatchTimeInterval = .seconds(2), systemMetricsType: SystemMetrics.Type? = nil) {
+        self.lock.withWriterLockVoid {
+            self.timeInterval = interval
+            if let type = systemMetricsType {
+                self.systemMetricsType = type
+            } else {
+                #if os(Linux)
+                self.systemMetricsType = LinuxSystemMetricsProvider.self
+                #else
+                self.systemMetricsType = UnsuportedMetricsProvider.self
+                #endif
+            }
+        }
+        self.updateSystemMetrics()
+    }
+
+    public static func cancelSystemMetrics() {
+        self.cancellable?.cancel()
+    }
+
+    fileprivate static func updateSystemMetrics() {
+        let interval = self.lock.withReaderLock { self.timeInterval }
+        let c = self.queue.schedule(after: .init(.now()), interval: .init(interval)) {
+            let prefix = "process_"
+            let pid = ProcessInfo.processInfo.processIdentifier
+            let metrics = self.systemMetricsType.init(pid: "\(pid)")
+            if let vmem = metrics.virtualMemory { Gauge(label: prefix + "virtual_memory_bytes").record(vmem) }
+            if let rss = metrics.residentMemory { Gauge(label: prefix + "resident_memory_bytes").record(rss) }
+            if let start = metrics.startTimeSeconds { Gauge(label: prefix + "start_time_seconds").record(start) }
+            if let cpuSeconds = metrics.cpuSeconds { Gauge(label: prefix + "cpu_seconds_total").record(cpuSeconds) }
+            if let maxFds = metrics.maxFds { Gauge(label: prefix + "max_fds").record(maxFds) }
+            if let openFds = metrics.openFds { Gauge(label: prefix + "open_fds").record(openFds) }
+        }
+        self.lock.withWriterLockVoid {
+            self.cancellable = c
+        }
+    }
+}
+
+public protocol SystemMetrics {
+    init(pid: String)
+
+    var virtualMemory: Int32? { get }
+    var residentMemory: Int32? { get }
+    var startTimeSeconds: Int32? { get }
+    var cpuSeconds: Int32? { get }
+    var maxFds: Int32? { get }
+    var openFds: Int32? { get }
+}
+
+#if os(Linux)
+private struct LinuxSystemMetricsProvider: SystemMetrics {
+    let virtualMemory: Int32?
+    let residentMemory: Int32?
+    let startTimeSeconds: Int32?
+    let cpuSeconds: Int32?
+    let maxFds: Int32?
+    let openFds: Int32?
+
+    private enum SystemMetricsError: Error {
+        case FileNotFound
+    }
+
+    init(pid: String) {
+        let ticks = Int32(_SC_CLK_TCK)
+        do {
+            guard let statString =
+                try String(contentsOfFile: "/proc/\(pid)/stat", encoding: .utf8)
+                .split(separator: ")")
+                .last
+            else { throw SystemMetricsError.FileNotFound }
+            let stats = String(statString)
+                .split(separator: " ")
+                .map(String.init)
+            self.virtualMemory = Int32(stats[safe: 20])
```

Comment:
not super keen on introducing one off `safe:` accessors like that, rather let's do the `if let value = ... { self.virtualMemory = value` usual dance?

Reply:
OR `?? ""` when needed?

Reply:
OR a form of `getIntStat(stats, 11)` etc?

Reply:
Reasoning for this is that normally an array does not return optionals or throws errors. It will crash if you go out of index, thus the `[safe:]` subscript that checks the bounds. I'm okay with moving this to a separate function, but don't fully see the point of that.